### PR TITLE
Implement global perms

### PR DESCRIPTION
**Pull request by Lulalaby**

### CODE REVIEW

File: DisCatSharp.ApplicationCommands/ApplicationCommandsExtension.cs [hunk 1]
```diff
@@ -24,6 +24,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Reflection;
+using System.Security;
```
Comment (Lulalaby) [inline DisCatSharp.ApplicationCommands/ApplicationCommandsExtension.cs:27]:
```suggestion
```

Rude

File: DisCatSharp.ApplicationCommands/ApplicationCommandsExtension.cs [hunk 2]
```diff
@@ -638,6 +637,24 @@ private void RegisterCommands(IEnumerable<ApplicationCommandsModuleConfiguration
...
+										foreach(var guild in perms)
+										{
+											guildOverwrites.Add(new KeyValuePair<ulong, DiscordGuildApplicationCommandPermission>(guild, new DiscordGuildApplicationCommandPermission()
+											{
+												Id = commandId.Value,
+												Permissions = permissions.Where(x => x.Key == guild).Select(x => x.Value).ToList(),
+												GuildId = guild
+											}));
+										}
+									}
```
Comment (JBraunsmaJr) [inline DisCatSharp.ApplicationCommands/ApplicationCommandsExtension.cs:656]: I notice this block of code is nearly identical to other spots. This feels like it should be a function of sorts

File: DisCatSharp.ApplicationCommands/ApplicationCommandsExtension.cs [hunk 3]
```diff
@@ -657,6 +674,24 @@ private void RegisterCommands(IEnumerable<ApplicationCommandsModuleConfiguration
...
+											foreach (var guild in perms)
+											{
+												guildOverwrites.Add(new KeyValuePair<ulong, DiscordGuildApplicationCommandPermission>(guild, new DiscordGuildApplicationCommandPermission()
+												{
+													Id = commandId.Value,
+													Permissions = permissions.Where(x => x.Key == guild).Select(x => x.Value).ToList(),
+													GuildId = guild
+												}));
+											}
+										}
```
Comment (JBraunsmaJr) [inline DisCatSharp.ApplicationCommands/ApplicationCommandsExtension.cs:693]: copy pasta could be function

File: DisCatSharp.ApplicationCommands/ApplicationCommandsExtension.cs [hunk 4]
```diff
@@ -680,6 +715,24 @@ private void RegisterCommands(IEnumerable<ApplicationCommandsModuleConfiguration
...
+												foreach (var guild in perms)
+												{
+													guildOverwrites.Add(new KeyValuePair<ulong, DiscordGuildApplicationCommandPermission>(guild, new DiscordGuildApplicationCommandPermission()
+													{
+														Id = commandId.Value,
+														Permissions = permissions.Where(x => x.Key == guild).Select(x => x.Value).ToList(),
+														GuildId = guild
+													}));
+												}
+											}
```
Comment (JBraunsmaJr) [inline DisCatSharp.ApplicationCommands/ApplicationCommandsExtension.cs:734]: copy pasta could be function

**Review by JBraunsmaJr — CHANGES_REQUESTED** [review]
We should try and reduce some of the copy-pasta going on. The only variation in these scenarios seems to be the parameters passed in in

File: DisCatSharp.ApplicationCommands/ApplicationCommandsExtension.cs (hunk 2 above)
Comment (KazutoKashima) [reply DisCatSharp.ApplicationCommands/ApplicationCommandsExtension.cs:656]: Agreed, we could cut it down to something like: (this is a mix of C++ and C#)

```
void OverwriteNewPerms([&bot], int* commandId, ptr::vector* permissions) {
    var (success, commandId, permissions) = PermissionsWorker.ResolveGlobalPermissions(types, command.id, com.Name, source.Value, source.Key);

    if (success) 
    {
          var perms = permissions.Select(x => x.Key).Distinct();
          foreach (var guild in perms)
          {
               // Here i cba writing the keyPair stuff
          }
     } else {
         Console.Write(/*the error code*/);
    }
};
```

Comment (Lulalaby) [reply DisCatSharp.ApplicationCommands/ApplicationCommandsExtension.cs:656]: hmm

**Pull request merged**